### PR TITLE
Feature/metrics instrumentation urllib3

### LINKED ISSUE #1144: Metrics instrumentation urllib3
HTPP metrics semconv https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/semantic_conventions/http-metrics.md

### PR DESCRIPTION
**Pull request by shalevr**
# Description
Add metrics instrumentation for urllib3

Fixes #1144 

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] tox -e test-instrumentation-urllib3

# Does This PR Require a Core Repo Change?

- [x] No.

# Checklist:

See [contributing.md](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/CONTRIBUTING.md) for styleguide, changelog guidelines, and more.

- [x] Followed the style guidelines of this project
- [x] Changelogs have been updated
- [x] Unit tests have been added

### CODE REVIEW

File: instrumentation/opentelemetry-instrumentation-urllib3/src/opentelemetry/instrumentation/urllib3/__init__.py [hunk 1]
```diff
@@ -135,8 +137,31 @@ def _instrument(self, **kwargs):
...
         tracer_provider = kwargs.get("tracer_provider")
         tracer = get_tracer(__name__, __version__, tracer_provider)
+
+        meter_provider = kwargs.get("meter_provider")
+        meter = get_meter(__name__, __version__, meter_provider)
+
+        duration_histogram = meter.create_histogram(
+            name="http.client.duration",
+            unit="ms",
+            description="measures the duration outbound HTTP requests",
```
Comment (lzchen) [inline instrumentation/opentelemetry-instrumentation-urllib3/src/opentelemetry/instrumentation/urllib3/__init__.py:147]:
```suggestion
            description="measures the duration of the outbound HTTP requests",
```

File: instrumentation/opentelemetry-instrumentation-urllib3/tests/test_urllib3_ip_support.py [hunk 2]
```diff
@@ -84,3 +84,50 @@ def assert_success_span(
...
+            resource_metrics = self.memory_metrics_reader.get_metrics_data().resource_metrics
+            for metrics in resource_metrics:
+                for scope_metrics in metrics.scope_metrics:
+                    self.assertEqual(len(scope_metrics.metrics), 3)
+                    for metric in scope_metrics.metrics:
+                        for data_point in metric.data.data_points:
+                            self.assertDictEqual(
+                                expected_attributes, dict(data_point.attributes)
+                            )
+                            self.assertEqual(data_point.count, 1)
```
Comment (lzchen) [inline instrumentation/opentelemetry-instrumentation-urllib3/tests/test_urllib3_ip_support.py:133]: Any way to test out the values?

Comment (shalevr) [reply instrumentation/opentelemetry-instrumentation-urllib3/tests/test_urllib3_ip_support.py:133]: assertDictEqual check if the expected_attributes dict is equal to the metrics attribues, the keys and the values

Comment (lzchen) [reply instrumentation/opentelemetry-instrumentation-urllib3/tests/test_urllib3_ip_support.py:133]: I meant testing out the data point values on the histogram.

Comment (shalevr) [reply instrumentation/opentelemetry-instrumentation-urllib3/tests/test_urllib3_ip_support.py:174]: Done

Comment (srikanthccv) [reply instrumentation/opentelemetry-instrumentation-urllib3/tests/test_urllib3_ip_support.py:167]: I think he is referring to the data point value, not the attribute value.

Comment (shalevr) [reply instrumentation/opentelemetry-instrumentation-urllib3/tests/test_urllib3_ip_support.py:167]: I verify the request and response data size like this:
 expected_data = {
                "http.client.request.size": len(body),
                "http.client.response.size": len(response.data)
            }
self.assertEqual(data_point.sum, expected_data[metric.name])

Comment (srikanthccv) [reply instrumentation/opentelemetry-instrumentation-urllib3/tests/test_urllib3_ip_support.py:167]: There is no test for values of duration

Comment (shalevr) [reply instrumentation/opentelemetry-instrumentation-urllib3/tests/test_urllib3_ip_support.py:182]: Fixed :)

**Review by lzchen — APPROVED** [review]

File: instrumentation/opentelemetry-instrumentation-urllib3/src/opentelemetry/instrumentation/urllib3/__init__.py [hunk 3]
```diff
@@ -175,11 +203,38 @@ def instrumented_urlopen(wrapped, instance, args, kwargs):
...
+                SpanAttributes.NET_PEER_PORT: instance.port,
+            }
+
+            version = getattr(response, "version")
+            if version:
+                metric_labels[SpanAttributes.HTTP_FLAVOR] = (
+                    "1.1" if version == 11 else "1.0"
+                )
+
+            request_size = 0 if body is None else len(body)
```
Comment (srikanthccv) [inline instrumentation/opentelemetry-instrumentation-urllib3/src/opentelemetry/instrumentation/urllib3/__init__.py:229]: Should you be using `Content-Length` from the response headers? Is all body data read in-memory? What if the response is compressed?

File: instrumentation/opentelemetry-instrumentation-urllib3/src/opentelemetry/instrumentation/urllib3/version.py [hunk 4]
```diff
@@ -12,4 +12,4 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-__version__ = "0.32b0"
+__version__ = "0.33b0"
```
Comment (srikanthccv) [inline instrumentation/opentelemetry-instrumentation-urllib3/src/opentelemetry/instrumentation/urllib3/version.py:15]: You don't need to change this.

File: instrumentation/opentelemetry-instrumentation-urllib3/tests/test_urllib3_ip_support.py [hunk 5]
```diff
@@ -84,3 +86,123 @@ def assert_success_span(
...
+                        for data_point in metric.data.data_points:
+                            if metric.name in expected_data:
+                                self.assertEqual(
+                                    data_point.sum, expected_data[metric.name]
+                                )
+                            if metric.name == "http.client.duration":
+                                self.assertAlmostEqual(
+                                    data_point.sum,
+                                    client_duration_estimated,
+                                    delta=5000,
```
Comment (srikanthccv) [inline instrumentation/opentelemetry-instrumentation-urllib3/tests/test_urllib3_ip_support.py:146]: This 5000ms delta is too high.

File: instrumentation/opentelemetry-instrumentation-urllib3/src/opentelemetry/instrumentation/urllib3/__init__.py [hunk 6]
```diff
@@ -175,11 +203,38 @@ def instrumented_urlopen(wrapped, instance, args, kwargs):
             inject(headers)
 
             with _suppress_further_instrumentation():
+                start_time = default_timer()
                 response = wrapped(*args, **kwargs)
+                elapsed_time = (default_timer() - start_time) * 1000
```
[the remark this answers is not in the document]
Comment (shalevr) [reply instrumentation/opentelemetry-instrumentation-urllib3/src/opentelemetry/instrumentation/urllib3/__init__.py]: Fixed

File: instrumentation/opentelemetry-instrumentation-urllib3/src/opentelemetry/instrumentation/urllib3/version.py (hunk 4 above)
Comment (shalevr) [reply instrumentation/opentelemetry-instrumentation-urllib3/src/opentelemetry/instrumentation/urllib3/version.py]: Ohh I didnt know... Done

File: instrumentation/opentelemetry-instrumentation-urllib3/src/opentelemetry/instrumentation/urllib3/__init__.py (hunk 3 above)
Comment (shalevr) [reply instrumentation/opentelemetry-instrumentation-urllib3/src/opentelemetry/instrumentation/urllib3/__init__.py:229]: Good point, I Changed only the response to get the Content-Length, in the request I cant get the Content-Length from the header

File: instrumentation/opentelemetry-instrumentation-urllib3/tests/test_urllib3_ip_support.py (hunk 5 above)
Comment (shalevr) [reply instrumentation/opentelemetry-instrumentation-urllib3/tests/test_urllib3_ip_support.py:146]: I changed it to 100 ms, what do you think?

Comment (srikanthccv) [reply instrumentation/opentelemetry-instrumentation-urllib3/tests/test_urllib3_ip_support.py:146]: I am fine with anything under somethings like 500ms or 1000ms to account for all cases and to make sure tests don't become flaky.

Comment (shalevr) [reply instrumentation/opentelemetry-instrumentation-urllib3/tests/test_urllib3_ip_support.py:147]: Done

**Review by srikanthccv — CHANGES_REQUESTED** [review]
I believe `uninstrument` should stop producing all telemetry including metrics. I am going to bring this up in today's SIG. Blocking the accidental merge meanwhile.

**Comment by lzchen** [comment]
Please update the metrics support for this instrumentation [here](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/instrumentation/README.md)

**Comment by srikanthccv** [comment]
Add `_supports_metrics=True` in the package.py file like [this](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/instrumentation/opentelemetry-instrumentation-requests/src/opentelemetry/instrumentation/requests/package.py#L18) and run `tox -e generate` and it should be updated automatically.

**Comment by srikanthccv** [comment]
@shalevr We will need tests to make sure `uninstrument` doesn't generate any measurements.

File: instrumentation/opentelemetry-instrumentation-urllib3/tests/test_urllib3_ip_support.py (hunk 5 above)
Comment (srikanthccv) [reply instrumentation/opentelemetry-instrumentation-urllib3/tests/test_urllib3_ip_support.py:160]: This doesn't seem to be addressed.

File: instrumentation/opentelemetry-instrumentation-urllib3/src/opentelemetry/instrumentation/urllib3/__init__.py [hunk 7]
```diff
@@ -254,6 +301,29 @@ def _is_instrumentation_suppressed() -> bool:
     )
 
 
+def create_metric_attributes(
```
[the remark this answers is not in the document]
Comment (shalevr) [reply instrumentation/opentelemetry-instrumentation-urllib3/src/opentelemetry/instrumentation/urllib3/__init__.py:304]: Fixed

**Review by srikanthccv — APPROVED** [review]

**Pull request merged**